### PR TITLE
feat: add repro for issue #4323 and cater for a conformity annotation

### PR DESCRIPTION
**Pull request by ggreif**
reproduce feature request #4323 and implement it

This won't restrict the resulting type ([as pointed out by @rossberg](https://github.com/dfinity/motoko/pull/4327#discussion_r1430238318)), but merely verifies the interface. `public` and type annotations on bindings _inside_ the block are there for shaping the former (just like in `class` definitions).

TODO:
- [x] documentation
- [x] `Changelog.md`
- [ ] resolve old-ish problem now marked with `FIXME` — #4355

### CODE REVIEW

File: doc/md/examples/grammar.txt [hunk 1]
```diff
@@ -284,15 +287,15 @@
...
 
 <dec_var> ::= 
-    'var' <id> (':' <typ>)? '=' <exp>
+    'var' <id> <annot_opt_eq> <exp>
 
 <dec_nonvar> ::= 
     'let' <pat> '=' <exp>
     'type' <id> ('<' <list(<typ_bind>, ',')> '>')? '=' <typ>
-    <obj_sort> <id>? '='? <obj_body>
+    <obj_sort> <id>? <annot_opt_eq>? <obj_body>
```
Comment (ggreif) [inline doc/md/examples/grammar.txt:298]:
```suggestion
    <obj_sort> <id>? '='? <obj_body>
```

revert this for now

**Comment by crusso** [comment]
Cool. Have you got a follow on PR? This one makes the grammar a bit weird (pulling the equal sign into the annotation grammar), so I'd rather review the effect of both in case there's a cleaner refactoring.

File: src/mo_frontend/parser.mly [hunk 2]
```diff
@@ -852,18 +854,22 @@ dec_nonvar :
       LetD (p', e', None) @? at $sloc }
   | TYPE x=typ_id tps=typ_params_opt EQ t=typ
     { TypD(x, tps, t) @? at $sloc }
-  | s=obj_sort xf=id_opt EQ? efs=obj_body
-    { let named, x = xf "object" $sloc in
+  | s=obj_sort xf=id_opt t=annot_opt_eq? efs=obj_body
+    { let t = Option.join t in
+      let sort = Type.(match s.it with
+                       | Actor -> "actor" | Module -> "module" | Object -> "object"
+                       | _ -> assert false) in
```
Comment (ggreif) [inline src/mo_frontend/parser.mly:861]: This covers `Memory`, but that cannot happen via `obj_sort`, so this is dead.

File: doc/md/basic-concepts.md [hunk 3]
```diff
@@ -150,7 +150,7 @@ x * y + x
 
 A declaration list is not itself (immediately) an *expression*, so we cannot (immediately) declare another variable with its final value (`3`).
 
-**Block expressions.** We can form a *block expression* from this list of declarations by enclosing it with matching *curly braces*. Blocks are only allowed as sub-expressions of control flow expressions like `if`, `loop`, `case`, etc. In all other places, we use `do { …​ }` to represent block expression, to distinguish blocks from object literals. For example, `do {}` is the empty block of type `()`, while `{}` is an empty record of record type `{}`.
+**Block expressions.** We can form a *block expression* from this list of declarations by enclosing it with matching *curly braces*. Blocks are only allowed as sub-expressions of control flow expressions like `if`, `loop`, `case`, etc. In all other places, we use `do { … }` to represent block expression, to distinguish blocks from object literals. For example, `do {}` is the empty block of type `()`, while `{}` is an empty record of record type `{}`.
```
Comment (ggreif) [inline doc/md/basic-concepts.md:153]: There was an invisible character in `do { … }`, I deleted it, maybe I shouldn't have!

File: doc/md/examples/grammar.txt [hunk 4]
```diff
@@ -292,7 +292,7 @@
 <dec_nonvar> ::= 
     'let' <pat> '=' <exp>
     'type' <id> ('<' <list(<typ_bind>, ',')> '>')? '=' <typ>
-    <obj_sort> <id>? '='? <obj_body>
+    <obj_sort> <id>? (':' <typ>)? '='? <obj_body>
```
Comment (ggreif) [inline doc/md/examples/grammar.txt:295]: In the chapter of `class` definitions, `<obj_body>` is said to contain the `=?`, but it is not consistent all over this document. This needs to be fixed-up.

Comment (ggreif) [reply doc/md/examples/grammar.txt:295]: I think it was a "hairline space" 0x200b. The `=?` problem was resolved in d1007d119.

File: doc/md/language-manual.md [hunk 5]
```diff
@@ -1600,6 +1602,8 @@ Evaluation of `<sort>? <id>? =? { <dec-field>;* }` proceeds by binding `<id>` (i
 
 If `<id>?` is present, the declaration binds `<id>` to `v`. Otherwise, it produces the empty set of bindings.
 
+If `(: <typ>)?` is present, the `<obj-body>` is checked against `<typ>`.
```
Comment (crusso) [inline doc/md/language-manual.md:1605]:
```suggestion
If `(: <typ>)?` is present, then `T` must be a subtype of  `<typ>`.
```

File: src/mo_frontend/typing.ml [hunk 6]
```diff
@@ -2434,7 +2445,7 @@ and infer_dec env dec : T.typ =
             display_typ_expand t'
             display_typ_expand t''
       | Some typ, T.Actor ->
-        local_error env dec.at "M0135" "actor class has non-async return type"
+         local_error env dec.at "M0135"(*FIXME: repeated?*) "actor class has non-async return type"
```
Comment (crusso) [inline src/mo_frontend/typing.ml:2448]: Fixme?

**Review by crusso — APPROVED** [review]
LGTM

File: src/mo_frontend/typing.ml (hunk 6 above)
Comment (ggreif) [reply src/mo_frontend/typing.ml:2448]: I'll try to fix this, though it is a previous problem.

**Comment by ggreif** [comment]
@mergifyIO refresh

**Comment by ggreif** [comment]
Closes #4327.

**Pull request merged**